### PR TITLE
[DEVX-433] Fix Ruby3 warnings

### PR DESCRIPTION
**Pull request by sveToptal**
Fix Ruby3 warnings

### Review

- [x] Document code according to [Getting Started with Yard](http://www.rubydoc.info/gems/yard/file/docs/GettingStarted.md).
- [x] All tests are passing.
- [x] Test manually.
- [x] Get approval.

### Pre-merge checklist

- [x] The PR relates to a single subject with a clear title and description in grammatically correct, complete sentences.
- [x] Verify that feature branch is up-to-date with `master` (if not - rebase it).
- [x] Double check the quality of [commit messages](http://chris.beams.io/posts/git-commit/).
- [x] Squash related commits together.

### CODE REVIEW

**Review by mrzasa — APPROVED** [review]

File: lib/granite/rspec/action_helpers.rb
Comment (sveToptal) [reply lib/granite/rspec/action_helpers.rb:4]: Frankly, I'm not sure of anything regarding Ruby3 compatibility, as I'm quite new to the epic :) 
I drew my "inspiration" from lines like [this](https://github.com/toptal/platform/blob/515be46fd70c1636b3927e25d539b93d07d5d300/app/decorators/talent_agreement_decorator.rb#L29) one in Platform and [this](https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/) article. Please let me know if I can improve it in any way.

Comment (mrzasa) [reply lib/granite/rspec/action_helpers.rb:4]: I'd say that there is no point in making this change and this should _just_ work with the previous syntax, but we need to test this.

**Review by mrzasa — COMMENTED** [review]
I'd re-check if we need the change to `delegate` syntax. Maybe it's just a false-positive?

File: lib/granite/rspec/action_helpers.rb
Comment (sveToptal) [reply lib/granite/rspec/action_helpers.rb:4]: I was able to find [similar cases](https://github.com/toptal/platform/pull/58292/files#diff-f046d40ca9b5e1878134a2866e21415e78f3f448d768fa6eaf49f4a145c93fe4R27) in Platform, which are already approved and merged. Not that it proves this being right, it rather shows a convention. I think it's mostly related to that. 

However, if you feel strongly about this, I can make some tests

Comment (sveToptal) [reply lib/granite/rspec/action_helpers.rb:4]: I've tested passing it like `{to: :subject}`. Test pass too. So I'd guess it's mostly a matter of preference and convention here.

Comment (mrzasa) [reply lib/granite/rspec/action_helpers.rb:4]: I don't have a strong opinion here, so let's wait for one more review :)

Comment (catks) [reply lib/granite/rspec/action_helpers.rb:4]: This really depends on how the method is implemented (delegate In this case), if it has `to:` as a keyword or only`**options` I think we can keep the current `delegate :perform!, to: :subject`

Which seems to be the case:

`def delegate(*methods, to: nil, prefix: nil, allow_nil: nil, private: nil)`([Link](https://apidock.com/rails/Module/delegate))

File: lib/granite/performer_proxy/proxy.rb
Comment (catks) [inline lib/granite/performer_proxy/proxy.rb:21]: Can't we just replace this by `(method, *args, **options, &block)`, to be compatible with ruby 2.6 instead of using the `ruby2_keywords`?
```suggestion
      ruby2_keywords def method_missing(method, *args, **options, &block)
        if @klass.respond_to?(method)
          @klass.with_proxy_performer(@performer) do
            @klass.public_send(method, *args, **options, &block)
```

Comment (sveToptal) [reply lib/granite/performer_proxy/proxy.rb:21]: @catks I tried your suggestion before coming up with this one. CI checks for ruby 2.6 were failing. That's why I decided to go with the current one.

File: lib/granite/rspec/action_helpers.rb
Comment (sveToptal) [reply lib/granite/rspec/action_helpers.rb:4]: oh, ok, I understand the reasoning now. Ok, good, let's keep it, but how should I disable the warning then?

Comment (catks) [reply lib/granite/rspec/action_helpers.rb:4]: I think we currently don't have a way of ignoring those in ruby3 checks yet, we could implement it, but I think adding a comment in the ticket/Platform PR is enough for the scope of this ticket.

You could also try out this section of the code on platform rails console to prove it works on ruby 3 (Or run a spec there with ruby 3 that uses this helper)

**Review by catks — APPROVED** [review]

File: lib/granite/rspec/action_helpers.rb
Comment (sveToptal) [reply lib/granite/rspec/action_helpers.rb]: tried it out, it's working ok. I reverted the old code. Will leave a comment in the ticket about this.

File: lib/granite/performer_proxy/proxy.rb
Comment (andriusch) [reply lib/granite/performer_proxy/proxy.rb:21]: Do you remember what the issues were? Perhaps there was something wrong with the test? Because I'm pretty sure 2.6 should support passing `**options` right? And we really only use this proxy for `as(performer)`, so keyword arguments shouldn't even matter.

Comment (catks) [reply lib/granite/performer_proxy/proxy.rb]: Ok, so this fixes the issue for ruby `2.7` but it still breaks ruby 2.6 since keywords work differently when using `send`, eg:

```ruby
# WIth ruby 2.6
def test(value) 
end

options = {}
test(42, **options) # works
send(:test, 42, **options) # raises ArgumentError (wrong number of arguments (given 2, expected 1))
```

It is possible to 'fix' some of the granite specs by changing the method signature (eg: https://github.com/toptal/granite/pull/73/commits/8eefbe74973c233dcb19ab5a33bd2eab4900f305#diff-1b00a5d80953e2e89aeaab3cfe7a7f7eafaf0d2a006305be4064284b7b28a8f0R10) but personally, I don't think it's a good approach as we will force ruby 2.6 users to also update their methods when upgrading.

Also [one of the specs](https://github.com/toptal/granite/blob/9f90ea83bf3d614e6ab17e6176d872507e77b730/spec/lib/granite/projector_spec.rb#L46) would need a different implementation for ruby 2.6 since an `Action` also receives `#new` with `[*params, {}]` instead of [*params].

Maybe we could remove the **ruby2_keywords** approach after dropping ruby 2.6 support or as another option we could  do something like this:

```ruby
   if RUBY_ENGINE && Gem::Version.new(RUBY_VERSION) >= Gem::Version.new("2.7")
        def method_missing(method, *args, **options, &block)
          if @klass.respond_to?(method)
            @klass.with_proxy_performer(@performer) do
              @klass.public_send(method, *args, **options, &block)
            end
          else
            super
          end
        end
      else
        def method_missing(method, *args, &block)
          if @klass.respond_to?(method)
            @klass.with_proxy_performer(@performer) do
              @klass.public_send(method, *args, &block)
            end
          else
            super
          end
        end
      end
```

But I guess I would keep the ruby2_keywords since it's simpler, WDYT? @andriusch

Comment (andriusch) [reply lib/granite/performer_proxy/proxy.rb:21]: OK let's keep `ruby2_keywords` and drop support for ruby 2.6 once `ruby2_keywords` is removed.

**Review by andriusch — APPROVED** [review]

**Pull request merged**
